// egg-ts-helper/dist/generators/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const d = require("debug");
const path = require("path");
const ts = require("typescript");
const utils = require("../utils");
const debug = d('egg-ts-helper#generators_config');
exports.EXPORT_DEFAULT_FUNCTION = 1;
exports.EXPORT_DEFAULT = 2;
exports.EXPORT = 3;
const cache = {};
function default_1(tsHelper) {
    tsHelper.register('config', (config, baseConfig) => {
        const fileList = utils.loadFiles(config.dir, config.pattern);
        const dist = path.resolve(config.dtsDir, 'index.d.ts');
        if (!fileList.length) {
            return { dist };
        }
        const importList = [];
        const declarationList = [];
        const moduleList = [];
        fileList.forEach(f => {
            const abUrl = path.resolve(config.dir, f);
            // read from cache
            if (!cache[abUrl] || config.file === abUrl) {
                f = f.substring(0, f.lastIndexOf('.'));
                const type = checkConfigReturnType(abUrl);
                const moduleName = utils.getModuleObjByPath(f).moduleName;
                const tsPath = path
                    .relative(config.dtsDir, path.join(config.dir, f))
                    .replace(/\/|\\/g, '/');
                debug('import %s from %s', moduleName, tsPath);
                const imn = `Export${moduleName}`;
                const prefix = type === exports.EXPORT ? '* as ' : '';
                const ims = `import ${prefix}${imn} from '${tsPath}';`;
                let tds = `type ${moduleName} = `;
                if (type === exports.EXPORT_DEFAULT_FUNCTION) {
                    tds += `ReturnType<typeof ${imn}>;`;
                }
                else if (type === exports.EXPORT_DEFAULT || type === exports.EXPORT) {
                    tds += `typeof ${imn};`;
                }
                else {
                    return;
                }
                // cache the file
                cache[abUrl] = {
                    import: ims,
                    declaration: tds,
                    moduleName,
                };
            }
            const cacheItem = cache[abUrl];
            importList.push(cacheItem.import);
            declarationList.push(cacheItem.declaration);
            moduleList.push(cacheItem.moduleName);
        });
        if (!importList.length) {
            return { dist };
        }
        const { base, inserts, property } = config.interface;
        const newType = `New${base}`;
        return {
            dist,
            content: `import { ${base} } from '${baseConfig.framework}';\n` +
                `${importList.join('\n')}\n` +
                `${declarationList.join('\n')}\n` +
                `type ${newType} = ${base} & ${moduleList.join(' & ')};\n\n` +
                `declare module '${baseConfig.framework}' {\n` +
                inserts
                    .map(prop => {
                    return (`  interface ${prop} {\n` +
                        `    ${property}: ${newType};\n` +
                        `  }\n`);
                })
                    .join('\n') +
                `}`,
        };
    });
}
exports.default = default_1;
// check config return type.
function checkConfigReturnType(f) {
    const sourceFile = utils.getSourceFile(f);
    if (!sourceFile) {
        return;
    }
    let hasExport = false;
    let exportElement;
    utils.eachSourceFile(sourceFile, node => {
        if (node.parent !== sourceFile) {
            return;
        }
        if (ts.isExportAssignment(node)) {
            // has export default ...
            exportElement = node.expression;
            return false;
        }
        else if (utils.modifierHas(node, ts.SyntaxKind.ExportKeyword)) {
            if (utils.modifierHas(node, ts.SyntaxKind.DefaultKeyword)) {
                exportElement = node;
                return;
            }
            // has export
            hasExport = true;
        }
    });
    if (exportElement) {
        return ts.isFunctionLike(exportElement)
            ? exports.EXPORT_DEFAULT_FUNCTION
            : exports.EXPORT_DEFAULT;
    }
    else if (hasExport) {
        return exports.EXPORT;
    }
}
exports.checkConfigReturnType = checkConfigReturnType;
