// egg-ts-helper/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const glob = require("globby");
const path = require("path");
const ts = require("typescript");
// load ts/js files
function loadFiles(cwd, pattern) {
    const fileList = glob.sync([pattern || '**/*.(js|ts)', '!**/*.d.ts'], {
        cwd,
    });
    return fileList.filter(f => {
        // filter same name js/ts
        return !(f.endsWith('.js') &&
            fileList.includes(f.substring(0, f.length - 2) + 'ts'));
    });
}
exports.loadFiles = loadFiles;
// clean same name js/ts
function cleanJs(cwd) {
    const fileList = [];
    glob
        .sync(['**/*.ts', '!**/*.d.ts', '!**/node_modules'], { cwd })
        .forEach(f => {
        const jf = removeSameNameJs(path.resolve(cwd, f));
        if (jf) {
            fileList.push(jf);
        }
    });
    if (fileList.length) {
        console.info(`[egg-ts-helper] These file was deleted because the same name ts file was exist!\n`);
        console.info('  ' + fileList.join('\n  ') + '\n');
    }
}
exports.cleanJs = cleanJs;
// get moduleName by file path
function getModuleObjByPath(f) {
    const props = f.split('/').map(prop => 
    // transfer _ to uppercase
    prop.replace(/[._-][a-z]/gi, s => s.substring(1).toUpperCase()));
    // composing moduleName
    const moduleName = props
        .map(prop => prop[0].toUpperCase() + prop.substring(1))
        .join('');
    return {
        props,
        moduleName,
    };
}
exports.getModuleObjByPath = getModuleObjByPath;
// remove same name js
function removeSameNameJs(f) {
    if (!f.endsWith('.ts') || f.endsWith('.d.ts')) {
        return;
    }
    const jf = f.substring(0, f.length - 2) + 'js';
    if (fs.existsSync(jf)) {
        fs.unlinkSync(jf);
        return jf;
    }
}
exports.removeSameNameJs = removeSameNameJs;
// parse ts file to ast
function getSourceFile(f) {
    const code = fs.readFileSync(f, {
        encoding: 'utf-8',
    });
    try {
        return ts.createSourceFile(f, code, ts.ScriptTarget.ES2017, true);
    }
    catch (e) {
        console.error(e);
        return;
    }
}
exports.getSourceFile = getSourceFile;
// check whether node was module.exports
function isModuleExports(node) {
    if (ts.isPropertyAccessExpression(node)) {
        const obj = node.expression;
        const prop = node.name;
        return (ts.isIdentifier(obj) &&
            obj.escapedText === 'module' &&
            ts.isIdentifier(prop) &&
            prop.escapedText === 'exports');
    }
    return false;
}
exports.isModuleExports = isModuleExports;
// check kind in node.modifiers.
function modifierHas(node, kind) {
    return node.modifiers && node.modifiers.find(mod => kind === mod.kind);
}
exports.modifierHas = modifierHas;
// each ast node
function eachSourceFile(node, cb) {
    const result = cb(node);
    if (result === false) {
        return;
    }
    node.forEachChild((sub) => {
        eachSourceFile(sub, cb);
    });
}
exports.eachSourceFile = eachSourceFile;
// require modules
function requireFile(url) {
    if (!fs.existsSync(url)) {
        return undefined;
    }
    let exp = require(url);
    if (exp.__esModule && 'default' in exp) {
        exp = exp.default;
    }
    if (typeof exp === 'function') {
        exp = exp();
    }
    return exp;
}
exports.requireFile = requireFile;
